modules/api_gateway/main.tf

```diff
@@ -0,0 +1,302 @@
+    }
+  }
+}
+
+data "aws_region" "current" {}
+
+# API GATEWAY AUTHORIZER LAMBDA
+data "aws_lambda_function" "lambda" {
+  for_each = var.lambdas
+
+  function_name = each.value
+}
+
+data "aws_iam_policy_document" "assume_role" {
+  statement {
+    effect  = "Allow"
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type = "Service"
+      identifiers = [
+        "apigateway.amazonaws.com",
+        "cloudwatch.amazonaws.com"
+      ]
+    }
+  }
+}
+
+resource "aws_iam_role" "invocation_role" {
+  name = "api_gateway_auth_invocation_role"
+  path = "/"
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "apigateway.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_api_gateway_authorizer" "authorizer" {
+  for_each = local.authorizer_keys
+
+  name        = "api_authorizer_${each.key}"
+  rest_api_id = aws_api_gateway_rest_api.rest_api.id
+  # authorizer_uri looks funny because of https://github.com/hashicorp/terraform-provider-aws/issues/26619
+  authorizer_uri                   = replace(data.aws_lambda_function.lambda[each.key].invoke_arn, "/\\:\\d{1,3}\\/invocations/", "/invocations")
+  authorizer_credentials           = aws_iam_role.invocation_role.arn
+  identity_source                  = "method.request.header.X-Auth-Token"
+  authorizer_result_ttl_in_seconds = 900
+}
+
+resource "aws_iam_role_policy" "invocation_policy" {
+  name = "apigateway_authorization_invocation_policy"
+  role = aws_iam_role.invocation_role.id
+
+  policy = jsonencode({
+    "Version" = "2012-10-17",
+    "Statement" = [
+      {
+        "Action" = "lambda:InvokeFunction",
+        "Effect" = "Allow",
+        "Resource" = [
+          for key in distinct(local.authorizer_list) : data.aws_lambda_function.lambda[key].arn
+        ]
+      }
+    ]
+  })
+}
+
+resource "aws_iam_role" "lambda_iam_role" {
+  name = "lambda_iam_role"
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "lambda.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+EOF
+}
+# END
+
+# API_GATEWAY
+resource "aws_lambda_permission" "web_rest_invoke_permission" {
+  for_each = local.integration_keys
+
+  statement_id  = "allow-${var.name}-restapi-invoke-${each.key}"
+  action        = "lambda:InvokeFunction"
+  function_name = data.aws_lambda_function.lambda[each.key].function_name
+  principal     = "apigateway.amazonaws.com"
+
+  # The /*/*/* part allows invocation from any stage, method and resource path
+  # within API Gateway REST API.
+  source_arn = "${aws_api_gateway_rest_api.rest_api.execution_arn}/*/*/*"
+}
+
+resource "aws_api_gateway_rest_api" "rest_api" {
+  name        = var.name
+  description = var.description
+
+  endpoint_configuration {
```

Comment:
I might make the type default to be "REGIONAL" but let it be configurable.  I would think more often then not we would have our own cloudfront distro that would route to the API gateway.

Reply:
I made "EDGE" the default because that is consistent with AWS's defaults. But it is configurable now.